VERSION_CATALOGS feature fails if any dir in project path starts with character "u": "illegal unicode escape"

```
FAILURE: Build failed with an exception.
* What went wrong:
org.gradle.api.internal.catalog.GeneratedClassCompilationException: Unable to compile generated sources:
  - File LibrariesForUnity.java, line: 31, illegal unicode escape
  - File LibrariesForUnity.java, line: 53, illegal unicode escape
> Unable to compile generated sources:
    - File LibrariesForUnity.java, line: 31, illegal unicode escape
    - File LibrariesForUnity.java, line: 53, illegal unicode escape
```

This is because "\u" is treated as a unicode escape even in comments:
`LibrariesForUnity.java` generated by gradle - note `\u` in the javadoc:
```java
...
        /**
         * Creates a dependency provider for platform (com.xxx:unity-xxx)
         * This dependency was declared in settings file '..\unityxxx\settings.gradle'
         */
        public Provider<MinimalExternalModuleDependency> getXxxxxx() { return create("xxxxxx"); }
...
```

This is happening when importing such project to IntelliJ IDEA version 2021.3.2 RC Build #IU-213.6777.38.

`unityxxx/settings.gradle`:
```groovy

include 'aaa', 'bbb'

enableFeaturePreview('VERSION_CATALOGS')
dependencyResolutionManagement {
    versionCatalogs {
        unity {
            version('unity', '3+')
            alias('p').to('com.xxx.xxx', 'unity-p').versionRef('unity')
        }
    }
}
```

`unityxxx/bbb/build.gradle`:
```groovy

dependencies {
    implementation unity.p
}
```

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/LibrariesSourceGenerator.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/LibrariesSourceGenerator.java
@@ -379,7 +379,7 @@ public class LibrariesSourceGenerator extends AbstractSourceGenerator {
         writeLn(" * If the version is a rich version and that its not expressible as a");
         writeLn(" * single version string, then an empty string is returned.");
         if (context != null) {
-            writeLn(" * This version was declared in " + context);
+            writeLn(" * This version was declared in " + sanitizeUnicodeEscapes(context));
         }
         writeLn(" */");
         String methodName = asProvider ? "asProvider" : "get" + toJavaName(leafNodeForAlias(versionAlias));
@@ -435,7 +435,7 @@ public class LibrariesSourceGenerator extends AbstractSourceGenerator {
         writeLn("    /**");
         writeLn("     * Creates a dependency provider for " + name + " (" + coordinates + ")");
         if (context != null) {
-            writeLn("     * This dependency was declared in " + context);
+            writeLn("     * This dependency was declared in " + sanitizeUnicodeEscapes(context));
         }
         writeLn("     */");
         String methodName = asProvider ? "asProvider": "get" + toJavaName(name);
@@ -468,7 +468,7 @@ public class LibrariesSourceGenerator extends AbstractSourceGenerator {
             }
             writeLn(" * </ul>");
             if (context != null) {
-                writeLn(" * This bundle was declared in " + context);
+                writeLn(" * This bundle was declared in " + sanitizeUnicodeEscapes(context));
             }
             writeLn(" */");
             String methodName = asProvider ? "asProvider" : "get" + toJavaName(leafNodeForAlias(alias));
@@ -482,13 +482,23 @@ public class LibrariesSourceGenerator extends AbstractSourceGenerator {
             writeLn("/**");
             writeLn(" * Creates a plugin provider for " + alias + " to the plugin id '" + id + "'");
             if (context != null) {
-                writeLn(" * This plugin was declared in " + context);
+                writeLn(" * This plugin was declared in " + sanitizeUnicodeEscapes(context));
             }
             writeLn(" */");
             String methodName = asProvider ? "asProvider" : "get" + toJavaName(leafNodeForAlias(alias));
             writeLn("public Provider<PluginDependency> " + methodName + "() { return createPlugin(\"" + alias + "\"); }");
         });
         writeLn();
+    }
+
+    /**
+     * Java compiler would fail to compile sources that have illegal unicode escape characters, including in the comments.
+     * Such characters could be accidentally introduced by a backslash followed by {@code 'u'},
+     * e.g. in Windows path {@code '..\\user\dir'}.
+     */
+    private static String sanitizeUnicodeEscapes(String s) {
+        // If a backslash precedes 'u', then we replace the backslash with its unicode notation '\\u005c'
+        return s.replace("\\u", "\\u005cu");
     }
 
     private static ClassNode rootNode(AccessorKind kind) {

--- a/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/catalog/LibrariesSourceGeneratorTest.groovy
+++ b/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/catalog/LibrariesSourceGeneratorTest.groovy
@@ -41,6 +41,7 @@ import org.gradle.process.ExecOperations
 import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
 import org.gradle.util.TestUtil
 import org.junit.Rule
+import spock.lang.Issue
 import spock.lang.Specification
 
 import java.util.function.Supplier
@@ -306,6 +307,30 @@ ${nameClash { noIntro().kind('dependency bundles').inConflict('one.cool', 'oneCo
         sources.hasBundle('b0Bundle', 'getB0Bundle', "This bundle was declared in ${context}")
         sources.hasVersion('v0Version', 'getV0Version', "This version was declared in ${innerContext}")
         sources.hasDependencyAlias('other', 'getOther', "This dependency was declared in ${context}")
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/19752")
+    def "backslashes are escaped when outputting context in javadocs"() {
+        def context = "Windows path: C:\\Users\\user\\Documents\\ultimate plugin"
+        def escapedContext = "Windows path: C:\\Users\\u005cuser\\Documents\\u005cultimate plugin"
+
+        when:
+        generate {
+            description.set("Some description for tests")
+            withContext(context) {
+                library("some-alias", 'g:a:v')
+                bundle("b0Bundle", ["some-alias"])
+                version("v0Version", "1.0")
+                library("other", "g", "a").versionRef("v0Version")
+            }
+        }
+
+        then:
+        sources.assertClass('Generated', 'Some description for tests')
+        sources.hasDependencyAlias('some-alias', 'getAlias', "This dependency was declared in ${escapedContext}")
+        sources.hasBundle('b0Bundle', 'getB0Bundle', "This bundle was declared in ${escapedContext}")
+        sources.hasVersion('v0Version', 'getV0Version', "This version was declared in ${escapedContext}")
+        sources.hasDependencyAlias('other', 'getOther', "This dependency was declared in ${escapedContext}")
     }
 
     private void generate(String className = 'Generated', @DelegatesTo(strategy = Closure.DELEGATE_FIRST, value = VersionCatalogBuilderInternal) Closure<Void> spec) {

